miner/worker.go

```diff
@@ -930,83 +999,57 @@ func (w *worker) commitNewWork(interrupt *int32, noempty bool, timestamp int64)
 			header.GasLimit = core.CalcGasLimit(parentGasLimit, w.config.GasCeil)
 		}
 	}
-	// Only set the coinbase if our consensus engine is running (avoid spurious block rewards)
-	if w.isRunning() {
-		if w.coinbase == (common.Address{}) {
-			log.Error("Refusing to mine without etherbase")
-			return
+	// Set the coinbase if the worker is running or it's required
+	if w.isRunning() || genParams.coinbase != (common.Address{}) {
+		if genParams.coinbase != (common.Address{}) {
+			header.Coinbase = genParams.coinbase
+		} else {
+			if w.coinbase == (common.Address{}) {
+				log.Error("Refusing to mine without etherbase")
+				return nil, errors.New("no etherbase specified")
+			}
+			header.Coinbase = w.coinbase
 		}
-		header.Coinbase = w.coinbase
 	}
+	// Run the consensus preparation with the default or customized consensus engine.
 	if err := w.engine.Prepare(w.chain, header); err != nil {
-		log.Error("Failed to prepare header for mining", "err", err)
-		return
-	}
-	// If we are care about TheDAO hard-fork check whether to override the extra-data or not
-	if daoBlock := w.chainConfig.DAOForkBlock; daoBlock != nil {
-		// Check whether the block is among the fork extra-override range
-		limit := new(big.Int).Add(daoBlock, params.DAOForkExtraRange)
-		if header.Number.Cmp(daoBlock) >= 0 && header.Number.Cmp(limit) < 0 {
-			// Depending whether we support or oppose the fork, override differently
-			if w.chainConfig.DAOForkSupport {
-				header.Extra = common.CopyBytes(params.DAOForkBlockExtra)
-			} else if bytes.Equal(header.Extra, params.DAOForkBlockExtra) {
-				header.Extra = []byte{} // If miner opposes, don't let it use the reserved extra-data
-			}
-		}
+		log.Error("Failed to prepare header for sealing", "err", err)
+		return nil, err
 	}
 	// Could potentially happen if starting to mine in an odd state.
-	err := w.makeCurrent(parent, header)
+	env, err := w.makeEnv(parent, header, w.coinbase)
```

Comment:
There's an ambiguity here. 

Up above, on line `1003`-`1011`, you prefer `genParams.coinbase`. If that is not set, you fall back to `w.coinbase`. However, here you use `w.coinbase` unconditionally. 

Is this an intentional feature, e.g. the ability to set the coinbase for a particular header but not change the 'global' coinbase? 

Otherwise, it seems like it might be better to require `genParams.coinbase` to be set, and use that consistently within this method.

Reply:
So the EL should have the ability to set the coinbase independent to the genParams.coinbase if it wants to. Don't know if that was the intention behind this change though. I think for most use cases we don't want to set the coinbase separately from the genParams.coinbase

Reply:
Updated. The coinbase can be passed by catalyst.